src/parquet/file/metadata-internal.cc

```diff
@@ -0,0 +1,525 @@
+  }
+
+  std::unique_ptr<RowGroupMetaData> GetRowGroupMetaData(int i) {
+    DCHECK(i < num_row_groups())
+        << "The file only has " << num_row_groups()
+        << " row groups, requested metadata for row group: " << i;
+    return RowGroupMetaData::GetRowGroupMetaData(
+        reinterpret_cast<const uint8_t*>(&metadata_->row_groups[i]));
+  }
+
+  const SchemaDescriptor* schema() const { return &schema_; }
+
+ private:
+  std::unique_ptr<const format::FileMetaData> metadata_;
+  SchemaDescriptor schema_;
+};
+
+std::unique_ptr<FileMetaData> FileMetaData::GetFileMetaData(const uint8_t* metadata) {
+  return std::unique_ptr<FileMetaData>(new FileMetaData(metadata));
+}
+
+FileMetaData::FileMetaData(const uint8_t* metadata)
+    : impl_{std::unique_ptr<FileMetaDataImpl>(new FileMetaDataImpl(metadata))} {}
+FileMetaData::~FileMetaData() {}
+
+std::unique_ptr<RowGroupMetaData> FileMetaData::GetRowGroupMetaData(int i) const {
+  return impl_->GetRowGroupMetaData(i);
+}
+
+int64_t FileMetaData::num_rows() const {
+  return impl_->num_rows();
+}
+
+int FileMetaData::num_row_groups() const {
+  return impl_->num_row_groups();
+}
+
+int32_t FileMetaData::version() const {
+  return impl_->version();
+}
+
+const std::string& FileMetaData::created_by() const {
+  return impl_->created_by();
+}
+
+int FileMetaData::num_schema_elements() const {
+  return impl_->num_schema_elements();
+}
+
+const SchemaDescriptor* FileMetaData::schema() const {
+  return impl_->schema();
+}
+
+void FileMetaData::WriteTo(OutputStream* dst) {
+  return impl_->WriteTo(dst);
+}
+
+// MetaData Builders
+// row-group metadata
+class ColumnMetaDataBuilder::ColumnMetaDataBuilderImpl {
+ public:
+  explicit ColumnMetaDataBuilderImpl(std::shared_ptr<WriterProperties> props,
+      const ColumnDescriptor* column, uint8_t* contents)
+      : properties_(props), column_(column) {
+    column_chunk_ = reinterpret_cast<format::ColumnChunk*>(contents);
+    column_chunk_->meta_data.__set_type(ToThrift(column->physical_type()));
+    column_chunk_->meta_data.__set_path_in_schema(column->path()->ToDotVector());
+    column_chunk_->meta_data.__set_codec(
+        ToThrift(properties_->compression(column->path())));
+    std::vector<format::Encoding::type> thrift_encodings;
+    std::set<Encoding::type> encodings_set;
+    ColumnEncodingSet encodings = properties_->encodings(column->path());
+    encodings_set.insert(encodings.def_level_encoding);
+    encodings_set.insert(encodings.rep_level_encoding);
+    if (properties_->dictionary_enabled()) {
+      encodings_set.insert(encodings.dictionary_encoding);
+    }
+    encodings_set.insert(encodings.value_encoding);
+    for (auto encoding : encodings_set) {
+      thrift_encodings.push_back(ToThrift(encoding));
+    }
+    column_chunk_->meta_data.__set_encodings(thrift_encodings);
+  }
+  ~ColumnMetaDataBuilderImpl() {}
+
+  // column chunk
+  void set_file_path(const std::string& val) { column_chunk_->__set_file_path(val); }
```

Comment:
Can you add a comment that this parameter is only used when a dataset is spread across multiple files? We are currently not conforming to the Parquet spec because we are not writing the column metadata after each column chunk. Creating a JIRA https://github.com/apache/parquet-format#file-format